src/gba/hardware-extensions.c

```diff
@@ -0,0 +1,339 @@
+/* Copyright (c) 2021 Jeffrey Pfau
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+#include <mgba/internal/gba/hardware-extensions.h>
+#include <mgba/internal/gba/memory.h>
+#include <mgba/internal/gba/io.h>
+#include <mgba/internal/gba/gba.h>
+
+
+enum HWEX_RETURN_CODES {
+    HWEX_RET_OK = 0,
+    HWEX_RET_WAIT = 1,
+
+    // Errors
+    HWEX_RET_ERR_UNKNOWN = 0x100,
+    HWEX_RET_ERR_BAD_ADDRESS = 0x101,
+    HWEX_RET_ERR_INVALID_PARAMETERS = 0x102,
+    HWEX_RET_ERR_WRITE_TO_ROM = 0x103,
+    HWEX_RET_ERR_ABORTED = 0x104,
+    HWEX_RET_ERR_DISABLED = 0x105,
+    HWEX_RET_ERR_DISABLED_BY_USER = 0x106
+};
+
+#define SIMPLIFY_HWEX_REG_ADDRESS(address) ((address - REG_HWEX0_CNT) >> 1)
+
+const uint16_t extensionIdByRegister[] = {
+    // More RAM
+    [SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_CNT)] = HWEX_ID_MORE_RAM,
+    [SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_RET_CODE)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P0_LO)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P0_HI)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P1_LO)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P1_HI)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P2_LO)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P2_HI)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P3_LO)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P3_HI)] = HWEX_ID_MORE_RAM,
+};
```

Comment:
Inconsistent tabs vs spaces (mGBA uses tabs)

Reply:
Fixed. I could swear I configured VS code to use tabs...